File: extra 1. String to Float(stof)/stof.py
__author__ = 'Bill Quan'


class Solution:
    def myStof(self, str):
        """

        :param str: str
        :return: float
        """
        str = str.strip()
        sign = +1
        if str[0] == '-':
            sign = -1
        if str[0] == '+':
            sign = +1
        i = 1
        rst = 0
        while i < len(str) and str[i].isdigit():
            rst = rst * 10 + int(str[i])  # ord(str[i]) - ord('0') is ok as well
            i += 1

        dec = 0
        if i < len(str) and str[i] == '.':
            i += 1
            div = 0.1
            while i < len(str) and str[i].isdigit():
                dec += int(str[i]) * div
                div *= 0.1
                i += 1

        return sign * (rst + dec)

# TEST:
test = Solution()
print test.myStof("1"), " || 1"
print test.myStof("-67"), " || -67"
print test.myStof("-67.890"), " || -67.89"
print test.myStof("  -0.00001234567890"), " || -0.0000123456789"
print test.myStof("  -0000.+00890"), " || 0"
print test.myStof("       -67.89"), " || -67.89"
print test.myStof("       ++67.89"), " || 0"
print test.myStof("       +n67.89"), " || 0"
print test.myStof("  +67.89sad"), " || 67.89"



"""
# overflow TEST
# the exponent of float in python is -1024 ~ +1023
t = 2.0 ** 1023
print t, type(t)
"""

File: extra 1. String to Float(stof)/stof.cpp
using namespace std;
#include <string>
#include <iostream>

class Solution
{
public:
	float myStof(string str)
	{
		int i = 0;
		int sign = 1;
		float rst = 0;
		while (i < str.length() && str[i] == ' ')
		{
			i++;
		}

		if (str[i] == '-')
		{
			sign = -1;
			i++;
		}
		else if (str[i] == '+')
		{
			i++;
		}

		while (i < str.length() && str[i] <= '9' && str[i] >= '0')
		{
			if (rst >= FLT_MAX / 10)
				return sign * FLT_MAX;
			rst = rst * 10 + str[i] - '0';
			i++;
		}

		float dec = 0;
		float exp = 0.1;
		if (i < str.length() && str[i] == '.')
		{
			i++;
			while (i < str.length() && str[i] <= '9' && str[i] >= '0')
			{
				dec += (str[i] - '0') * exp;
				exp *= 0.1;
				i++;
			}
		}
		return sign * (rst + dec);
	}
};

int main()
{
	Solution sol;
	float a = FLT_MIN;
	cout << sol.myStof("   -0.1234567890");

	
	system("pause");
	return 0;
}